core/tauri/src/ipc/protocol.rs

```diff
@@ -0,0 +1,275 @@
+                options: message.options,
+              })
+            }),
+        );
+      }
+    }
+
+    match invoke_message
+      .unwrap_or_else(|| serde_json::from_str::<Message>(&message).map_err(Into::into))
+    {
+      Ok(message) => {
+        let _ = window.on_message(InvokeRequest {
+          cmd: message.cmd,
+          callback: message.callback,
+          error: message.error,
+          body: message.payload.into(),
+          headers: message.options.map(|o| o.headers.0).unwrap_or_default(),
+        });
+      }
+      Err(e) => {
+        let _ = window.eval(&format!(
+          r#"console.error({})"#,
+          serde_json::Value::String(e.to_string())
+        ));
+      }
+    }
+  }
+}
+
+fn handle_ipc_request<R: Runtime>(
+  request: &HttpRequest,
+  manager: &WindowManager<R>,
+  label: &str,
+) -> std::result::Result<InvokeResponse, String> {
+  if let Some(window) = manager.get_window(label) {
+    // TODO: consume instead
+    #[allow(unused_mut)]
+    let mut body = request.body().clone();
+
+    let cmd = request
+      .uri()
+      .strip_prefix("ipc://localhost/")
+      .map(|c| c.to_string())
+      // the `strip_prefix` only returns None when a request is made to `https://tauri.$P` on Windows
+      // where `$P` is not `localhost/*`
+      // in this case the IPC call is considered invalid
+      .unwrap_or_else(|| "".to_string());
+    let cmd = percent_encoding::percent_decode(cmd.as_bytes())
+      .decode_utf8_lossy()
+      .to_string();
+
+    // the body is not set if ipc_custom_protocol is not enabled so we'll just ignore it
+    #[cfg(all(feature = "isolation", ipc_custom_protocol))]
+    if let crate::Pattern::Isolation { crypto_keys, .. } = manager.pattern() {
+      body = crate::utils::pattern::isolation::RawIsolationPayload::try_from(&body)
+        .and_then(|raw| crypto_keys.decrypt(raw))
+        .map_err(|e| e.to_string())?;
+    }
+
+    let callback = CallbackFn(
+      request
+        .headers()
+        .get(TAURI_CALLBACK_HEADER_NAME)
+        .ok_or("missing Tauri-Callback header")?
+        .to_str()
+        .map_err(|_| "Tauri callback header value must be a string")?
+        .parse()
+        .map_err(|_| "Tauri callback header value must be a numeric string")?,
+    );
+    let error = CallbackFn(
+      request
+        .headers()
+        .get(TAURI_ERROR_HEADER_NAME)
+        .ok_or("missing Tauri-Error header")?
+        .to_str()
+        .map_err(|_| "Tauri error header value must be a string")?
+        .parse()
+        .map_err(|_| "Tauri error header value must be a numeric string")?,
+    );
+
+    let content_type = request
+      .headers()
+      .get(reqwest::header::CONTENT_TYPE)
+      .and_then(|h| h.to_str().ok())
+      .map(|mime| mime.parse())
+      .unwrap_or(Ok(mime::APPLICATION_OCTET_STREAM))
+      .map_err(|_| "unknown content type")?;
+    let body = if content_type == mime::APPLICATION_OCTET_STREAM {
```

Comment:
I think we can turn this back into a match statement. I'll submit a change

Reply:
Since the value is manually implemented instead of derived, we cannot use the constant as the pattern